source/common/router/route_config_update_receiver_impl.cc

```diff
@@ -78,10 +81,12 @@ bool RouteConfigUpdateReceiverImpl::onVhdsUpdate(
     const VirtualHostRefVector& added_vhosts, const std::set<std::string>& added_resource_ids,
     const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
-
-  auto vhosts_after_this_update =
-      std::make_unique<std::map<std::string, envoy::config::route::v3::VirtualHost>>(
-          *vhds_virtual_hosts_);
+  std::unique_ptr<VirtualHostMap> vhosts_after_this_update;
+  if (vhds_virtual_hosts_ != nullptr) {
+    vhosts_after_this_update = std::make_unique<VirtualHostMap>(*vhds_virtual_hosts_);
+  } else {
+    vhosts_after_this_update = std::make_unique<VirtualHostMap>();
+  }
```

Comment:
naive question: could you skip making the empty map the calls to removeHosts and updateVhosts if vhds_virtual_hosts is null?

just set 'removed' and 'updated' to 0?

I admit I didn't read deeply here so maybe those functions have more side effects.

Reply:
I don't really want to get into changing onVhdsUpdate here, just trying to improve onRdsUpdate :)